rfc/0020-group-policies.md

```diff
@@ -0,0 +1,263 @@
+
+## Controller and CRDs
+
+This RFC proposes to add two new CRDs to the Kubewarden controller:
+
+- `GroupAdmissionPolicy`
+- `GroupClusterAdmissionPolicy`
+
+The CRDs will share most of the fields with the `AdmissionPolicy` and `ClusterAdmissionPolicy` CRDs, respectively.
+The main difference is that the `GroupAdmissionPolicy` and `GroupClusterAdmissionPolicy` CRDs will have a list of policies.
+Also, the `GroupAdmissionPolicy` and `GroupClusterAdmissionPolicy` CRDs will have a field to specify the boolean expression that will be used to evaluate the policies
+and a `message` field to specify the message that will be returned when the group policy is rejected.
+
+## Example CRD
+
+```yaml
+apiVersion: policies.kubewarden.io/v1
+kind: ClusterAdmissionPolicyGroup # or AdmissionPolicyGroup
+metadata:
+  name: pod-image-signatures
+spec:
+  rules:
+    - apiGroups: [""]
+      apiVersions: ["v1"]
+      resources: ["pods"]
+      operations:
+        - CREATE
+        - UPDATE
+  policies:
+    - name: sigstore_pgp
+      module: ghcr.io/kubewarden/policies/verify-image-signatures:v0.2.8
+      settings:
+        signatures:
+          - image: "*"
+            pubKeys:
+              - "-----BEGIN PUBLIC KEY-----xxxxx-----END PUBLIC KEY-----"
+              - "-----BEGIN PUBLIC KEY-----xxxxx-----END PUBLIC KEY-----"
+    - name: sigstore_gh_action
+      module: ghcr.io/kubewarden/policies/verify-image-signatures:v0.2.8
+      settings:
+        signatures:
+          - image: "*"
+            githubActions:
+            owner: "kubewarden"
+    - name: reject_latest_tag
+      module: ghcr.io/kubewarden/policies/trusted-repos-policy:v0.1.12
+      settings:
+        tags:
+          reject:
+            - latest
+  expression: "sigstore_pgp() || (sigstore_gh_action() && reject_latest_tag())"
+  message: "The group policy is rejected."
+```
+
+### Expression language
+
+We will use [CEL](https://github.com/google/cel-go) as the expression language for the group policies.
+The main reason for this choice is that CEL is used by the [ValidatingAdmissionPolicy](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/) and [matchConditions](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-matchconditions) in Kubernetes.
+
+Each policy in the group will be represented as a function call in the expression with the same name as the policy defined in the group.
+The expression field should be a valid CEL expression that evaluates to a boolean value and it will be validated by the Kubewarden controller's webhook.
+If the expression evaluates to `true`, the group policy will be considered as `accepted`, otherwise, it will be considered as `rejected`.
+
+### Message
+
+The message field will be used to specify the message that will be returned when the group policy is rejected.
+The specific policy results will be returned in the `warning` field of the response.
```

Comment:
Instead of warning messages, I would prefer to populate the admissionReview response.status, as it has  an array `statusDetails` that seems fitting: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.30/#statusdetails-v1-meta.

Reply:
I would prefer to return results using this approach.

Reply:
bump, just in case it was lost in the thread.

Reply:
@viccuad I haven't but we can have a discussion about this during the Parking lot today so I will leave this unresolved.

Reply:
It appears that `statusDetails.causes` is designed to return errors. See usage here: https://github.com/googleforgames/agones/blob/541dafe5684db40a169cdffdd5a16a4023260c20/pkg/util/webhooks/webhooks.go#L103 . 
We want instead to provide a list of information about the policy evaluation result. 
As VAP is using the warning fields for the failed expression, I think we should go with this one. (see https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/#validation-actions)

Reply:
Good find. Indeed, I agree, let's go with the warning fields. Resolving.